internal/jsonutil/reader.go

```diff
@@ -0,0 +1,126 @@
+package jsonutil
+
+import (
+	"encoding/json"
+	"errors"
+	"fmt"
+	"io"
+
+	"iter"
+)
+
+// StreamArrayReader reads a JSON array from r and yields each element.
+// keys is a list of keys hierarchy to traverse before reading the array.
+func StreamArrayReader[T any](r io.Reader, keys []string) iter.Seq2[T, error] {
```

Comment:
According to the docs it might be good to add a note here that this is a "single-use iterator" https://pkg.go.dev/iter@master#hdr-Single_Use_Iterators